csharp/src/Google.Protobuf/MessageExtensions.cs

```diff
@@ -145,6 +146,39 @@ public static ByteString ToByteString(this IMessage message)
             return ByteString.AttachBytes(message.ToByteArray());
         }
 
+        /// <summary>
+        /// Writes the given message data to the given buffer writer in protobuf encoding.
+        /// </summary>
+        /// <param name="message">The message to write to the stream.</param>
+        /// <param name="output">The stream to write to.</param>
+        public static void WriteTo(this IMessage message, IBufferWriter<byte> output)
+        {
+            ProtoPreconditions.CheckNotNull(message, nameof(message));
+            ProtoPreconditions.CheckNotNull(output, nameof(output));
+
+            WriteContext.Initialize(output, out WriteContext ctx);
+            WritingPrimitivesMessages.WriteRawMessage(ref ctx, message);
+            ctx.Flush();
+
+            // TODO: handling errors when IBufferWriter is used?
```

Comment:
@JamesNK  what is the best pattern for handling serialization exceptions while writing to an IBufferWriter?
I'm talking about a situation where I've got a buffer using writer.GetSpan() and then in the middle of the serialization I get a protobuf serialization error. It sounds like I might need a `finally` statement that invalidates the current buffer segment by calling `Advance(current position)`(otherwise there's a buffer leak potentially?). Or is this supposed to be solved on the higher level (i.e. protobuf serialization fails, so the owner of the IBufferWriter invalidates the entire IBufferWriter which also releases the current buffer segment
)?:. I'd expect there would be a best practice to follow in this case.

Reply:
You don't need to do anything at this level.

For example:
1. You call GetSpan
2. An error occurs and is caught while serialization is still happening
3. Serialization resumes and GetSpan is called again

The second GetSpan call means the previous `Span<byte>` buffer is no longer active. (the second GetSpan could have returned the same segment of buffer memory, or it might not)

Or for example:
1. You call GetSpan
2. An error occurs and is not caught
3. The PipeWriter (which implements `IBufferWriter<byte>`) is marked as complete without Advance ever being called.

The buffer reserved by GetSpan is thrown away.